tests/testthat/test-recall_data.R

```diff
@@ -0,0 +1,206 @@
+    expect_equal(returned_dtypes, expected_dtypes)
+})
+
+test_that("recall_by_model has the correct shape",{
+    result <- recall_by_model('Civic', limit=25, partial=TRUE)
+    expect_type(result, "list")
+    expect_equal(nrow(result), 25)
+    expected_col_names <- c("Recall number","Manufacturer Name",
+                            "Model name", "Make name", "Year","Recall date")
+    expect_equal(colnames(result), expected_col_names)
+    expected_col_names <- c("Recall number","Manufacturer Name",
+                            "Model name", "Make name", "Year","Recall date")
+    expect_equal(colnames(result), expected_col_names)
+    expected_dtypes <- c("character", "character", "character",
+                         "character", "integer",   "Date")
+    returned_dtypes <- as.vector(unlist(sapply(result, class)))
+    expect_equal(returned_dtypes, expected_dtypes)
+    result <- recall_by_model('Civic', start_year = 1980)
+    expect_gt(nrow(result), 0)
+    result <- recall_by_model('Civic', end_year = 1980)
+    expect_gt(nrow(result), 0)
+})
+
+test_that("recall_by_model by year range has the correct shape",{
+    result <- recall_by_model('Civic', start_year = 1980, end_year=2005)
+    expect_gt(nrow(result), 0)
+    filtered <- subset(result, Year >= 1980 & Year <= 2005)
+    expect_equal(nrow(result), nrow(filtered))
+    expected_col_names <- c("Recall number","Manufacturer Name",
+                            "Model name", "Make name", "Year","Recall date")
+    expect_equal(colnames(result), expected_col_names)
+    expected_dtypes <- c("character", "character", "character",
+                         "character", "integer",   "Date")
+    returned_dtypes <- as.vector(unlist(sapply(result, class)))
+    expect_equal(returned_dtypes, expected_dtypes)
+})
+
+test_that("recall_by_model using partial has the correct shape",{
+    result <- recall_by_model('Civic', partial = TRUE)
+    names <- unique(result$`Model name`)
+    expect_gt(length(names), 1)
+    result <- recall_by_model('Civic', partial = FALSE)
+    names <- unique(result$`Model name`)
+    expect_equal(length(names), 1)
+    expected_col_names <- c("Recall number","Manufacturer Name",
+                            "Model name", "Make name", "Year","Recall date")
+    expect_equal(colnames(result), expected_col_names)
+    expected_dtypes <- c("character", "character", "character",
+                         "character", "integer",   "Date")
+    returned_dtypes <- as.vector(unlist(sapply(result, class)))
+    expect_equal(returned_dtypes, expected_dtypes)
+})
+
+test_that("recall_by_year returns the correct shape",{
+    result <- recall_by_years(2010, 2012)
+    expect_type(result, "list")
+    expect_equal(nrow(result), 25)
+    expected_col_names <- c("Recall number","Manufacturer Name",
+                            "Model name", "Make name", "Year","Recall date")
+    expect_equal(colnames(result), expected_col_names)
```

Comment:
all of these calls to recall are really similar between the various functions tested, changes in the output would result in a fair amount of work rewriting tests, might be worth considering using some of the user defined functions suggested in R packages

Reply:
yeah ill read up on r test udfs

Reply:
we decided that doing that (UDF test suite) is outside the scope of this PR